File: packages/protocol/src/envelope.ts
/**
 * Signaling envelope (spec section 6). Every control-plane message between two devices,
 * over the LAN endpoint or relayed by the rendezvous server, is one of these.
 */
import { b64urlDecode, b64urlEncode, utf8Decode, utf8Encode } from './encoding.ts';
import { ProtocolError } from './errors.ts';
import { sign, verify, type Identity } from './identity.ts';

export const PROTOCOL_VERSION = 1;
export const SUPPORTED_VERSIONS: readonly number[] = [1];
export const SIGNALING_CONTEXT = 'prc-signaling-v1';
export const MAX_ENVELOPE_BYTES = 65536;
export const MAX_CLOCK_SKEW_MS = 300_000;

export const SIGNALING_TYPES = [
  'PAIR_REQUEST',
  'PAIR_RESULT',
  'SESSION_REQUEST',
  'SESSION_CHALLENGE',
  'SESSION_AUTH',
  'SESSION_ACCEPT',
  'SESSION_REJECT',
  'SDP_OFFER',
  'SDP_ANSWER',
  'ICE_CANDIDATE',
  'SESSION_RESUME',
  'SESSION_END',
] as const;
export type SignalingType = (typeof SIGNALING_TYPES)[number];

export function isSignalingType(t: string): t is SignalingType {
  return (SIGNALING_TYPES as readonly string[]).includes(t);
}

export interface UnsignedEnvelope {
  v: number;
  type: string;
  from: string;
  to: string;
  /** Empty string before a session exists (pairing, SESSION_REQUEST). */
  session: string;
  /** Starts at 1 and increases by 1 per sender per session. */
  seq: number;
  /** Unix milliseconds. */
  ts: number;
  /** base64url of the exact UTF-8 JSON bytes of the payload object. */
  payload: string;
}

export interface Envelope extends UnsignedEnvelope {
  sig: string;
}

/** The bytes that are signed. Fields joined by newline after the context label. */
export function signingInput(e: UnsignedEnvelope): Uint8Array {
  const s = [SIGNALING_CONTEXT, String(e.v), e.type, e.from, e.to, e.session, String(e.seq), String(e.ts), e.payload].join('\n');
  return utf8Encode(s);
}

export function encodePayload(value: unknown): string {
  return b64urlEncode(utf8Encode(JSON.stringify(value)));
}

export function decodePayload(payload: string): unknown {
  try {
    return JSON.parse(utf8Decode(b64urlDecode(payload)));
  } catch {
    throw new ProtocolError('invalid_payload');
  }
}

export async function signEnvelope(unsigned: UnsignedEnvelope, privateKey: CryptoKey): Promise<Envelope> {
  const sig = await sign(privateKey, signingInput(unsigned));
  return { ...unsigned, sig: b64urlEncode(sig) };
}

export async function verifyEnvelopeSignature(e: Envelope, publicKey: CryptoKey): Promise<boolean> {
  let sig: Uint8Array;
  try {
    sig = b64urlDecode(e.sig);
  } catch {
    return false;
  }
  const { sig: _sig, ...unsigned } = e;
  return verify(publicKey, signingInput(unsigned), sig);
}

export function serializeEnvelope(e: Envelope): string {
  return JSON.stringify(e);
}

/** Builds and signs outgoing envelopes, tracking `seq` per (recipient, session). */
export class EnvelopeSender {
  readonly #identity: Identity;
  readonly #seq = new Map<string, number>();
  readonly #now: () => number;

  constructor(identity: Identity, now: () => number = () => Date.now()) {
    this.#identity = identity;
    this.#now = now;
  }

  get deviceId(): string {
    return this.#identity.deviceId;
  }

  async build(type: SignalingType, to: string, session: string, payload: unknown): Promise<Envelope> {
    const key = `${to}|${session}`;
    const seq = (this.#seq.get(key) ?? 0) + 1;
    this.#seq.set(key, seq);
    return signEnvelope(
      { v: PROTOCOL_VERSION, type, from: this.#identity.deviceId, to, session, seq, ts: this.#now(), payload: encodePayload(payload) },
      this.#identity.privateKey,
    );
  }

  forgetSession(to: string, session: string): void {
    this.#seq.delete(`${to}|${session}`);
  }
}

File: apps/android/app/src/main/kotlin/com/prc/controller/protocol/Envelope.kt
package com.prc.controller.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * The signaling envelope (spec section 6). Every control-plane message between two devices is one
 * of these, signed by the sender over a fixed joined string rather than over the JSON, so no
 * canonical JSON form is needed anywhere.
 */
@Serializable
data class Envelope(
    val v: Int,
    val type: String,
    val from: String,
    val to: String,
    val session: String,
    val seq: Long,
    val ts: Long,
    val payload: String,
    val sig: String,
) {
    companion object {
        const val PROTOCOL_VERSION = 1
        const val SIGNALING_CONTEXT = "prc-signaling-v1"
        const val MAX_ENVELOPE_BYTES = 65536
        const val MAX_CLOCK_SKEW_MS = 300_000L
        val SUPPORTED_VERSIONS = listOf(1)

        val TYPES = setOf(
            "PAIR_REQUEST", "PAIR_RESULT", "SESSION_REQUEST", "SESSION_CHALLENGE", "SESSION_AUTH",
            "SESSION_ACCEPT", "SESSION_REJECT", "SDP_OFFER", "SDP_ANSWER", "ICE_CANDIDATE",
            "SESSION_RESUME", "SESSION_END",
        )

        val json = Json { ignoreUnknownKeys = false; encodeDefaults = true; explicitNulls = true }

        /** The bytes that are signed: the context label and the fields, joined by newlines. */
        fun signingInput(
            v: Int, type: String, from: String, to: String, session: String, seq: Long, ts: Long, payload: String,
        ): ByteArray = Encoding.utf8(
            listOf(SIGNALING_CONTEXT, v.toString(), type, from, to, session, seq.toString(), ts.toString(), payload)
                .joinToString("\n")
        )

        fun encodePayload(json: String): String = Encoding.b64url(Encoding.utf8(json))
    }

    fun signingInput(): ByteArray = signingInput(v, type, from, to, session, seq, ts, payload)

    fun payloadJson(): String = Encoding.utf8Decode(Encoding.b64urlDecode(payload))

    fun serialize(): String = json.encodeToString(serializer(), this)
}

/** Builds and signs outgoing envelopes, counting seq per recipient and session. */
class EnvelopeSender(
    private val identity: SigningIdentity,
    private val now: () -> Long = { System.currentTimeMillis() },
) {
    private val seq = HashMap<String, Long>()

    val deviceId: String get() = identity.deviceId

    fun build(type: String, to: String, session: String, payloadJson: String): Envelope {
        val key = "$to|$session"
        val next = (seq[key] ?: 0L) + 1
        seq[key] = next
        val payload = Envelope.encodePayload(payloadJson)
        val input = Envelope.signingInput(
            Envelope.PROTOCOL_VERSION, type, identity.deviceId, to, session, next, now(), payload,
        )
        // Rebuild with the same timestamp the signature covers.
        val ts = readTimestamp(input)
        val sig = Encoding.b64url(identity.sign(input))
        return Envelope(Envelope.PROTOCOL_VERSION, type, identity.deviceId, to, session, next, ts, payload, sig)
    }

    /** The timestamp is inside the signed string, so it is read back rather than sampled twice. */
    private fun readTimestamp(input: ByteArray): Long =
        Encoding.utf8Decode(input).split("\n")[7].toLong()

    /** A retry outside any session is a fresh attempt, and the host resets its own counter too. */
    fun resetAttempt(to: String) {
        seq.remove("$to|")
    }

    fun forgetSession(to: String, session: String) {
        seq.remove("$to|$session")
    }
}
